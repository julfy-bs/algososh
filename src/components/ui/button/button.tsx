import { FC, HTMLProps } from 'react';
import styles from './button.module.css';
import loaderIcon from '../../../images/icons/loader.svg';
import { AscendingIcon } from '../icons/ascending-icon';
import { DescendingIcon } from '../icons/descending-icon';
import { Direction } from '../../../types/sort';
import { clsx } from 'clsx';

type ButtonProps = {
  text?: string;
  type?: 'button' | 'submit' | 'reset';
  sorting?: Direction;
  linkedList?: 'small' | 'big';
  isLoader?: boolean;
  extraClass?: string;
} & HTMLProps<HTMLButtonElement>

export const Button: FC<ButtonProps> = ({
  text,
  extraClass = '',
  type = 'button',
  isLoader = false,
  sorting,
  linkedList,
  disabled,
  ...rest
}) => {
  const currentIcon =
    sorting === Direction.Ascending
      ? <AscendingIcon />
      : <DescendingIcon />;

  const className = clsx(
    'text',
    'text_type_button',
    'text_color_primary',
    styles.button,
    extraClass,
    {
      [styles.loader]: isLoader,
      [styles.small]: linkedList === 'small',
      [styles.big]: linkedList === 'big',
    });

  return (
    <button
      className={ className }
      type={ type }
      disabled={ isLoader || disabled }
      { ...rest }
    >
      { isLoader
        ? (
          <img
            className={ clsx(styles.loader_icon) }
            src={ loaderIcon }
            alt="Загрузка."
          />
        )
        : (
          <>
            { sorting && currentIcon }
            <p className={ clsx(text, { ['ml-5']: sorting }) }>{ text }</p>
          </>
        ) }
    </button>
  );
};
