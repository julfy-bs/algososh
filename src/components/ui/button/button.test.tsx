import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Button } from './button';

describe('Компонент <Button /> рендерится без ошибок', () => {
  it('с текстом', () => {
    const mockString: string = 'Button';
    const button = renderer
      .create(<Button text={ mockString } />)
      .toJSON();
    expect(button).toMatchSnapshot();
  });
  it('без текста', () => {
    const button = renderer
      .create(<Button />)
      .toJSON();
    expect(button).toMatchSnapshot();
  });
  it('с заблокированной кнопкой', () => {
    const button = renderer
      .create(<Button disabled={ true } />)
      .toJSON();
    expect(button).toMatchSnapshot();
  });
  it('с индикацией загрузки', () => {
    const button = renderer
      .create(<Button isLoader={ true } />)
      .toJSON();
    expect(button).toMatchSnapshot();
  });
});

describe('Компонент <Button /> кликабельный', () => {
  it('Нажатие на кнопку вызывает корректный alert', () => {
    window.alert = jest.fn();
    render(
      <Button
        text={ 'Кнопка' }
        onClick={ () => alert('Кнопка работает') }
      />
    );
    const button = screen.getByText('Кнопка');
    fireEvent.click(button);
    expect(window.alert).toHaveBeenCalledWith('Кнопка работает');
  });
});
