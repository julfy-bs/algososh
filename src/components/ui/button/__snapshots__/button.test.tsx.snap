// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when hovered 1`] = `undefined`;

exports[`changes the class when hovered 2`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  type="button"
>
  <p
    className="text undefined"
  />
</button>
`;

exports[`Компонент <Button /> рендерится без ошибок без текста 1`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  type="button"
>
  <p
    className="text undefined"
  />
</button>
`;

exports[`Компонент <Button /> рендерится без ошибок с заблокированной кнопкой 1`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  disabled={true}
  type="button"
>
  <p
    className="text undefined"
  />
</button>
`;

exports[`Компонент <Button /> рендерится без ошибок с индикацией загрузки 1`] = `
<button
  className="text text_type_button text_color_primary button undefined loader "
  disabled={true}
  type="button"
>
  <img
    alt="Загрузка."
    className="loader_icon"
    src="loader.svg"
  />
</button>
`;

exports[`Компонент <Button /> рендерится без ошибок с текстом 1`] = `
<button
  className="text text_type_button text_color_primary button undefined false "
  type="button"
>
  <p
    className="text undefined"
  >
    Button
  </p>
</button>
`;
